libp2p/muxers/mplex/coder.nim

```diff
@@ -50,26 +50,32 @@ proc writeMsg*(conn: Connection,
   trace "sending data over mplex", id,
                                   msgType,
                                   data = data.len
-  var
+  try:
+    var
       left = data.len
       offset = 0
-  while left > 0 or data.len == 0:
-    let
-      chunkSize = if left > MaxMsgSize: MaxMsgSize - 64 else: left
-      chunk = if chunkSize > 0 : data[offset..(offset + chunkSize - 1)] else: data
-    ## write lenght prefixed
-    var buf = initVBuffer()
-    buf.writePBVarint(id shl 3 or ord(msgType).uint64)
-    buf.writePBVarint(chunkSize.uint64) # size should be always sent
-    buf.finish()
-    left = left - chunkSize
-    offset = offset + chunkSize
-    try:
+    while left > 0 or data.len == 0:
+      let
+        chunkSize = if left > MaxMsgSize: MaxMsgSize - 64 else: left
+        chunk = if chunkSize > 0 : data[offset..(offset + chunkSize - 1)] else: data
+      ## write lenght prefixed
+      var buf = initVBuffer()
+      buf.writePBVarint(id shl 3 or ord(msgType).uint64)
+      buf.writePBVarint(chunkSize.uint64) # size should be always sent
+      buf.finish()
+      left = left - chunkSize
+      offset = offset + chunkSize
       await conn.write(buf.buffer & chunk)
-    except LPStreamIncompleteError as exc:
-      trace "unable to send message", exc = exc.msg
-    if data.len == 0:
-      return
+
+      if data.len == 0:
+        return
+  except LPStreamEOFError:
+    trace "Ignoring EOF while writing"
+  except CatchableError as exc:
```

Comment:
This doesn't look like a good place to swallow `CancelledError`.

Judging by the rest of the PR, I guess the motivation is that `write` errors can be ignored. This still doesn't sounds like a good reason to interrupt the cancellation process, because the write may be issued as a sub-step in the cancelled operation. It's better if the `CancelledError` propagates all the way to the original place where the operation was started.

cc @cheatfate .

Reply:
Agree with @zah, if you going to swallow exceptions like you do, please add one more handler for CancelledError before handler of `CatchableError`.
```nim
  except CancelledError as exc:
    raise exc
  except CatchableError as exc:
    ...
```

Reply:
pushed, here and elsewhere.

This is not a long-term plan perhaps, but right now, it's hard to tell where these exceptions end up and they're causing the client to crash. to the best of my estimation, it's happening during stream closure, but there are async races around that are hard to diagnose, and above all, hard to find, because there's quite a few layers involved. 

In particular, the hunch is that the nbc crashes are due to "internal" libp2p traffic and not requests originating from nbc itself - it's in part related to a race condition and connection ownership issue when closing connections that is being worked on - once that is solved, these exceptions here can be cleaned up as  well